Support service processes

jobhh opened the pull request:
Fixes #24 

What was done per commit:
1. Move the PhoenixActivity code into a separate (Kotlin) class.
2. Migrate ProcessPhoenix class to Kotlin.
3. Add triggerRebirth(Context, Class<?>) with Service rebirth support and triggerServiceRebirth(Context, Intent).
4. Change from Service to IntentService and remove NotificationBuilder from library.
5. Some cleanup.
6. Added Service support to the existing triggerRebirth(context, vararg nextIntents) function.

The ProcessPhoenix class is now Kotlin, but all public functions are annotated with @JvmStatic.
If this is undesired I can implement it in Java, but since Kotlin is my preference I started like this.

I changed from Service to IntentService (in commit 4) because this also seemed to be functional and IntentService's are meant to be short-lived, which the PhoenixService certainly is.

The sample has been updated to allow a easy test of this functionality.

**One important side-node (which is also mentioned in the Service classes):**
Restarting a Service multiple times will result in an increasingly long delay between restart times. Android seems to register the restart of this service as a crashed service and probably doesn't want to infinitely restart a crashing service, so that makes sense.
The increasing delay periods seem to follow the pattern: 4^x, with x being the restart attempt minus 1. I observed 1s, 4s, 16s, 64s, 256s and 1024s on the Zebra Android 11 device i used to test.

JakeWharton commented:
Thanks, but I would prefer to keep the library written in Java. Believe me, I love Kotlin as much as the next person, but this library doesn't really need it and I'd rather not bother with it at this point. Your change is the biggest the library has seen in like a decade, and I doubt anyone will add anything for another decade once this lands.

Commit:
Move PhoenixActivity to a separate class

Commit:
Add support for restarting a Service.

jobhh commented:
Alrighty, Java it is :)

Any feedback?

JakeWharton commented:
I didn't review the implementation, but will today.

jobhh commented:
Cool, no hurry. Let me know if anything needs to change.

JakeWharton commented on process-phoenix/src/main/java/com/jakewharton/processphoenix/PhoenixActivity.java:
```diff
@@ -0,0 +1,35 @@
+import android.os.Process;
+import android.os.StrictMode;
+
+public class PhoenixActivity extends Activity {
```
```suggestion
public final class PhoenixActivity extends Activity {
```

JakeWharton commented on process-phoenix/src/main/java/com/jakewharton/processphoenix/PhoenixService.java:
```diff
@@ -0,0 +1,54 @@
+ * The observed delay periods are: 1s, 4s, 16s, 64s, 256s, 1024s. (on an Android 11 device)
+ * Which seems to follow this pattern: 4^x, with x being the restart attempt minus 1.
+ */
+public class PhoenixService extends IntentService {
```
```suggestion
public final class PhoenixService extends IntentService {
```

JakeWharton commented on process-phoenix/src/main/java/com/jakewharton/processphoenix/ProcessPhoenix.java:
```diff
@@ -53,25 +55,83 @@ public static void triggerRebirth(Context context) {
+   * <p>
+   * Behavior of the current process after invoking this method is undefined.
+   */
+  public static void triggerRebirth(Context context, Class<?> targetClass) {
```
I don't really like this from an API design standpoint. I cannot imagine the case where you have either an activity or a service and need to call some general-purpose method to recreate it. Chances are you know what you have at the call-site. Someone that actually doesn't know can write the assignable check themselves.

If we want to offer `Class`-accepting methods then we can overload the specific methods with `Class<? extends Activity>` and `Class<? extends Service>`.

jobhh replied:
I based this on your input here: https://github.com/JakeWharton/ProcessPhoenix/issues/24#issuecomment-299610704

But, I do agree :) The caller should know whether an Activity or Service is to be restarted when they call this method.

JakeWharton replied:
Ahaha ouch. Bad 7-year-ago Jake!

JakeWharton commented on sample/src/main/AndroidManifest.xml:
```diff
@@ -18,5 +21,11 @@
+    <service
+      android:name=".RestartService"
+      android:exported="true"
+      android:foregroundServiceType="shortService" />
```
```suggestion
      android:foregroundServiceType="shortService"
      />
```

JakeWharton commented on sample/src/main/java/com/jakewharton/processphoenix/sample/NotificationBuilder.java:
```diff
@@ -0,0 +1,52 @@
+import android.os.Build;
+import android.util.Log;
+
+public class NotificationBuilder {
```
```suggestion
public final class NotificationBuilder {
```
Java is SO tedious

JakeWharton commented on sample/src/main/java/com/jakewharton/processphoenix/sample/RestartService.java:
```diff
@@ -0,0 +1,54 @@
+ * The observed delay periods are: 1s, 4s, 16s, 64s, 256s, 1024s. (on an Android 11 device)
+ * Which seems to follow this pattern: 4^x, with x being the restart attempt minus 1.
+ */
+public class RestartService extends IntentService {
```
```suggestion
public final class RestartService extends IntentService {
```

JakeWharton commented on sample/src/main/java/com/jakewharton/processphoenix/sample/RestartService.java:
```diff
@@ -0,0 +1,54 @@
+        // Trigger rebirth from a separate thread, such that the onStartCommand can finish properly
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.execute(() -> {
+            try {
+                Thread.sleep(1000);
+            } catch (InterruptedException e) {
+                // Ignore
+            }
```
```suggestion
            SystemClock.sleep(1000)
```

jobhh committed:
Apply suggestions from code review

Co-authored-by: Jake Wharton <github@jakewharton.com>

Commit:
Split Activity and Service methods, keeping the existing methods Activity-only.

Commit:
Merge branch 'service-restart' into trunk

# Conflicts:
#	process-phoenix/src/main/java/com/jakewharton/processphoenix/ProcessPhoenix.java
#	sample/src/main/java/com/jakewharton/processphoenix/sample/MainActivity.java

jobhh commented:
@JakeWharton The triggerRebirth(Context context, Intent... nextIntents) method already existed in previous versions, but it also added an Activity vs Service check in this pull request.

I've now removed these checks and opted to keep the previously existing methods Activity-only and add methods like triggerServiceRebirth(Context context, Intent nextIntent) for Service restarting.

Would this be ok?

Commit:
Run spotless code formatting
